### DISCUSSION TITLE
Creating a popup to a jointJs element added in a functional component not working

### DISCUSSION THREAD

**Discussion by Anuthama-Ganesan** [clientIO/joint#1825] [Q&A] [answered]
I have been trying to add a popup to a jointJs element (cell) on hover. As per the documentation, the target property of ui.popup in jointjs is recommended to point to "this". How do I do the same if my paper is initialised in a functional component? Pointing to that cell doesn't seem to work.

**Comment by jamesgeorgewilliams** [comment, marked as answer]
The `target` element needs to be present in the DOM in order for a Popup to be visible. The target can be any HTML or SVG.

Using a `paper` event, a popup could be displayed on hover:

```js
paper.on('element:mouseover', function(elementView) {
    (new joint.ui.Popup({
        content: '<b>I am a ui.Popup</b>',
        target: elementView.el
    })).render();
});
```

You can provide some code examples or more context if that doesn't help with your issue. Thanks

**Reply by Anuthama-Ganesan** [reply]
I tried the same, by pointing the target as "elementView.el", but the position of the content is not dynamically adjusting/positioning with respect to the target cell. As in, it is rendering in the same left and top positions. Should we manually alter the transforms? Because as far as I understand the popup should be rendering pointing to that target element? 
Also, I don't see a 'popup' component as shown in the documentation example- 
![image](https://user-images.githubusercontent.com/71628906/194234687-f901e098-9f6c-42a4-82af-c5708d5792f6.png)
Is there any specific component or should I have to enclose it in a div and still see a popup?

These are the code snippets used -
```
function onPaperCellMouserOver(_cellView, evt) {
    if (!_cellView.model.isLink() && _cellView.model.attributes.attrs.label.fill == 'white') {
      (new ui.Popup({ content: '<b>I am a ui.Popup</b>', target: _cellView.el })).render();
}

paper.on('element:mouseover', function (cellView, evt) {
      onPaperCellMouserOver(cellView, evt);
    });
```

**Reply by jamesgeorgewilliams** [reply]
The position of the popup is calculated from the target element. It should appear underneath your target element if everything is setup correctly, and you shouldn't need to modify `top` and `left` positions. 

Only one popup will be open at a time, and the `top` and `left` positions should change dynamically depending on the position of the target.

Your code snippet is working fine for me.  In the function above `onPaperCellMouseOver`, you are missing a closing curly bracket. `}`. I'm not sure if that is just a typo when you were copying the code.
```js
function onPaperCellMouserOver(_cellView, evt) {
    if (!_cellView.model.isLink() && _cellView.model.attributes.attrs.label.fill == 'blue') {
        (new joint.ui.Popup({ content: '<b>I am a ui.Popup</b>', target: _cellView.el })).render();
    }
}

paper.on('element:mouseover', function(cellView, evt) {
    onPaperCellMouserOver(cellView, evt);
});
```

I assume you are using JointJS+ if using the `ui.Popup`, you should also make sure your application is using the `rappid.css` file. If you are experiencing some problems such as the popup not displaying, that could be the root of the issue too.

**Reply by Anuthama-Ganesan** [reply]
Thanks for the clarification. I am using JointJS+ `ui.Popup`. I have installed the package using `rappid.tgz` file. But I am guessing that I am missing the rappid.css file, and that might be the root of the issue. I will add that and give it a try.

And to clarify, the closing brace is a typo.

**Reply by jamesgeorgewilliams** [reply]
After installing the package, the `rappid.css` file will be located in `node_modules/@clientio/rappid/rappid.css`.

If you are using webpack, you can import it in your css/scss as we have done here - `@import "~@clientio/rappid/rappid.css";`
https://github.com/clientIO/joint-plus-tutorial-react/blob/main/src/App.scss

If you are using another build tool like `vite`, you will need to adjust the import accordingly.

**Reply by Anuthama-Ganesan** [reply]
That seems to do the trick. Can we add custom styling to the content in the popup, say font style and colour?

**Reply by jamesgeorgewilliams** [reply]
You can add custom styling to the popup via CSS:

```css
.joint-popup {
       color: tomato;
       font-family: Arial, Helvetica, sans-serif;
}
```
